src/checkers/inference/solver/frontend/Lattice.java

```diff
@@ -70,20 +57,38 @@
      */
     public final int numTypes;
 
+    /**
+     * Runtime Annotation Mirrors extracted from slots.
+     * This field is useful for type systems that has a dynamic number
+     * of type qualifiers.
+     */
+    public final Collection<AnnotationMirror> runtimeAMs;
+
+    /**
+     * Underlying qualifier hierarchy that this lattice built based on.
+     * This field is nullable, it will be null if this lattice doesn't built based on
+     * a real qualifier hierarchy. (E.g. TwoQualifierLattice).
+     */
+    /* @Nullable */
+    private final QualifierHierarchy underlyingQualifierHierarchy;
+
     public Lattice(Map<AnnotationMirror, Collection<AnnotationMirror>> subType,
             Map<AnnotationMirror, Collection<AnnotationMirror>> superType,
             Map<AnnotationMirror, Collection<AnnotationMirror>> incomparableType,
-            Map<AnnotationMirror, Integer> typeToInt, Map<Integer, AnnotationMirror> intToType,
             Set<? extends AnnotationMirror> allTypes, AnnotationMirror top, AnnotationMirror bottom,
-            int numTypes) {
+            int numTypes, Collection<AnnotationMirror> runtimeAMs, /* @Nullable */ QualifierHierarchy qualifierHierarchy) {
```

Comment:
CF no longer uses annotations in comments ( /* @Nullable */  ==> @Nullable ). We can check to see if CFI still requires annotations in comments.

Reply:
Ok. After refactor Lattice class hierarchy, I think I will remove this `@nullable` annotation and make `qualifierHierarchy` always non-null.